src/url-loader/url-loader.ts

```diff
@@ -2,16 +2,37 @@
 /**
  * An object that reads files.
  */
-export interface UrlLoader {
+export abstract class UrlLoader {
   /**
    * Returns `true` if this loader can load the given `url`.
    */
-  canLoad(url: string): boolean;
+  abstract canLoad(url: string): boolean;
 
   /**
    * Reads a file from `url`.
    *
    * This should only be called if `canLoad` returns `true` for `url`.
    */
-  load(url: string): Promise<string>;
+  abstract load(url: string): Promise<string>;
+
+  /**
+   * Return true iff the loader is able to list nearby resources and give
+   * autocompletion help with a user that's typing out a path.
+   *
+   * A filesystem loader e.g. can do this pretty easily with fs.listdir, but
+   * a loader that fetches remote resouces may not be able to.
+   */
+  offersCompletions() {
+    return false;
+  }
+
+  /**
+   * Given the partial path (e.g. that a user may be in the process of typing)
+   * of `pathSoFar`, return an array of paths that pathSoFar could complete to.
+   *
+   * `pathSoFar` is a resolved URL relative to the package root.
```

Comment:
Confession: I think there's a number of bugs that will come up the first time we use a different url loader. I'd like to chat with you when you get a chance because package relative paths are an assumption that's tacitly made in a bunch of places, and I'm not sure how to think about url resolvers in general.